packages/atclient/src/connection.c

```diff
@@ -402,11 +403,55 @@ int atclient_connection_send(atclient_connection *ctx, const unsigned char *src,
   // if return value is positive and less than src_len
   // we should continue to write from the appropriate offset
   // (multiple writes must be summed to determine total data written)
-  if ((ret = mbedtls_ssl_write(&(ctx->ssl), src, src_len)) <= 0) {
-    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "mbedtls_ssl_write failed with exit code: %d\n", ret);
+  if (src[src_len - 1] != '\n') {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_WARN, "command does not have a trailing \\n character:\t%s\n", src);
+  }
+  if ((ret = mbedtls_ssl_write(&ctx->ssl, src, src_len)) <= 0) { // error only when the returned value is negative
+    mbedtls_strerror(ret, error_buf, sizeof(error_buf));
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "mbedtls_ssl_write returned -0x%x: %s\n", -ret, error_buf);
     goto exit;
   }
 
+  /*
+   * 4 a. In some cases, mbedtls_ssl_write returns a non-zero value. When only part of the data written.
+   * Repeat mbedtls_ssl_write until the ret is equal to remaining_length (which is the condition for successful write)
+   */
+  if (ret != src_len) {
+    int length_written = ret;
+    int remaining_length = src_len - length_written;
+    int attempt = 0;
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_DEBUG,
+                 "Partial write successful. Performing recursive write. Current ret: %d\n", ret);
+    // warning: potiential infinite loop
```

Comment:
I could not find the existing impl for this. So, for now I have reverted this change as part of [9e57229](https://github.com/atsign-foundation/at_c/pull/423/commits/9e57229bbdbd7f1ec1a8fe0590500925bfa4e110). And since this is not causing any issues yet, we can implement this as part of a different ticket in the future.